# Simulate unobserved sites 

You can use `-explode 1` to expand simulations to the sites not observed in the input VCF file.


Example command:
```shell
./vcfgl -i example.vcf -e 0.01 -d 5 -O v -d 3 --seed 42 -explode 1
```

Example input: `example.vcf`

Our input file contains two sites, and its header contains one chromosome named "chr22" with length 5.
```
##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##contig=<ID=chr22,length=5>
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	ind1	ind2
chr22	1	.	0	1	.	PASS	.	GT	0|0	0|0
chr22	3	.	0	1	.	PASS	.	GT	1|1	1|1
```


Example output: `output.vcf`

The missing 3 sites (at position 2,4, and 5) are generated by `-explode 1` by setting all individuals' genotypes to `0|0`. Thus the output file contains information for all sites.
```
##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##contig=<ID=chr22,length=5>
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##fileDate=Tue Sep 12 15:46:49 2023
##source=vcfgl --input example.vcf --output output --output-mode v --error-rate 0.010000 --depth 3.000000 --depths-file (null) --seed 42 -explode 1 -printBaseCounts 0 -addGP 0 -addPL 0 -addI16 0 -addQS 0
##source=vcfgl version: v0.3.0 ee51b23-dirty
##FORMAT=<ID=GL,Number=G,Type=Float,Description="Genotype likelihood in log10 likelihood ratio format">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Simulated per-sample read depth">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	ind1	ind2
chr22	1	.	A	C,G,T	.	PASS	.	GT:DP:GL	0|0:2:0,-0.59914,-4.94551,-0.59914,-4.94551,-4.94551,-0.59914,-4.94551,-4.94551,-4.94551	0|0:5:0,-1.49785,-12.3638,-1.49785,-12.3638,-12.3638,-1.49785,-12.3638,-12.3638,-12.3638
chr22	2	.	A	C,G,T	.	PASS	.	GT:DP:GL	0|0:3:0,-0.898711,-7.41827,-0.898711,-7.41827,-7.41827,-0.898711,-7.41827,-7.41827,-7.41827	0|0:2:0,-0.59914,-4.94551,-0.59914,-4.94551,-4.94551,-0.59914,-4.94551,-4.94551,-4.94551
chr22	3	.	A	C,G,T	.	PASS	.	GT:DP:GL	1|1:1:-2.47276,-0.29957,0,-2.47276,-0.29957,-2.47276,-2.47276,-0.29957,-2.47276,-2.47276	1|1:4:-9.89103,-1.19828,0,-9.89103,-1.19828,-9.89103,-9.89103,-1.19828,-9.89103,-9.89103
chr22	4	.	A	C,G,T	.	PASS	.	GT:DP:GL	0|0:5:0,-1.49785,-12.3638,-1.49785,-12.3638,-12.3638,-1.49785,-12.3638,-12.3638,-12.3638	0|0:2:0,-0.59914,-4.94551,-0.59914,-4.94551,-4.94551,-0.59914,-4.94551,-4.94551,-4.94551
chr22	5	.	A	C,G,T	.	PASS	.	GT:DP:GL	0|0:4:0,-1.19828,-9.89103,-1.19828,-9.89103,-9.89103,-1.19828,-9.89103,-9.89103,-9.89103	0|0:4:0,-1.19828,-9.89103,-1.19828,-9.89103,-9.89103,-1.19828,-9.89103,-9.89103,-9.89103
```
